File: 03_pub_sub/src/pub_sub_cpp/src/subscriber.cpp
#include <memory> // 智能指针

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

// 自定义订阅者节点，继承 rclcpp::Node
class FirstSubscriber : public rclcpp::Node
{
public:
    FirstSubscriber()
        : Node("subscriber_cpp")
    {
        // 定义话题消息回调函数
        // 当订阅者接收到消息时，该函数会被调用
        auto topic_callback =
            [this](std_msgs::msg::String::UniquePtr msg) -> void
        {
            // 输出接收到的消息内容
            RCLCPP_INFO(
                this->get_logger(),
                "I heard: '%s'",
                msg->data.c_str());
        };

        // 创建订阅者：
        // 消息类型为 std_msgs::msg::String
        // 话题名为 cpp_topic
        // QoS 队列深度为 10
        // 接收到消息后调用 topic_callback
        subscription_ =
            this->create_subscription<std_msgs::msg::String>(
                "cpp_topic",
                10,
                topic_callback);
    }

private:
    // String 类型的话题订阅者
    rclcpp::Subscription<std_msgs::msg::String>::SharedPtr subscription_;
};

int main(int argc, char *argv[])
{
    // 初始化 ROS 2 C++ 客户端库
    rclcpp::init(argc, argv);

    // 创建 FirstSubscriber 节点并进入事件循环，
    // 等待并处理订阅消息
    rclcpp::spin(std::make_shared<FirstSubscriber>());

    // 关闭 ROS 2 C++ 客户端库
    rclcpp::shutdown();

    return 0;
}

File: 03_pub_sub/src/pub_sub_python/pub_sub_python/subscriber.py
import rclpy
from rclpy.node import Node

from std_msgs.msg import String

# 定义订阅者节点类
class FirstSubscriber(Node):
    def __init__(self):
        # 创建名为 subscriber_python 的节点
        super().__init__('subscriber_python')

        # 创建订阅者：
        # 消息类型为 String，订阅 python_topic 话题
        # 接收到消息后调用 subscriber_callback()
        # 队列深度为 10
        self.subscriber = self.create_subscription(
            String,
            'python_topic',
            self.subscriber_callback,
            10
        )

    # 订阅回调函数，每接收到一条消息就会被调用一次
    def subscriber_callback(self, msg):
        self.get_logger().info(f'接收到消息，内容为：{msg.data}')

def main(args=None):
    # 初始化 ROS 2 Python 客户端库
    rclpy.init(args=args)

    # 创建订阅者节点
    subscriber_node = FirstSubscriber()

    # 输出节点启动日志
    subscriber_node.get_logger().info(
        '已创建 subscriber_python，准备接收消息...'
    )

    # 保持节点运行，等待并处理接收到的消息
    rclpy.spin(subscriber_node)

    # 关闭 ROS 2 Python 客户端库
    rclpy.shutdown()

if __name__ == '__main__':
    main()
